File: java/revenj-core/src/main/java/org/revenj/security/UserPrincipal.java
package org.revenj.security;

import java.security.Principal;
import java.util.Set;
import java.util.function.Function;

public final class UserPrincipal implements Principal {

	private final String name;
	private final Function<String, Boolean> isInRole;

	public UserPrincipal(String name, Function<String, Boolean> isInRole) {
		this.name = name;
		this.isInRole = isInRole;
	}

	public UserPrincipal(String name, Set<String> roles) {
		this.name = name;
		this.isInRole = roles::contains;
	}

	@Override
	public String getName() {
		return name;
	}

	public boolean implies(String role) {
		return name.equals(role) || isInRole.apply(role);
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/security/UserPrincipal.scala
package net.revenj.security

import java.security.Principal

class UserPrincipal(name: String, isInRole: String => Boolean) extends Principal {

  def this(name: String, roles: Set[String]) {
    this(name, r => roles.contains(r))
  }

  override def getName: String = name

  def implies(role: String): Boolean = name == role || isInRole.apply(role)
}

